### PR TITLE
Make `-(::BigFloat)` inherit precision of input

### LINKED ISSUE #52100: Inappropriate number of digits displayed by `show(::Complex{BigFloat}`
While the precisions of the imaginary parts of  `c` and `d` are identical, the negative one is
displayed with more decimal digits, wrongly suggesting higher accuracy.

```
julia> d = 1 + BigFloat(pi, precision=32)*im
1.0 + 3.1415926535im

julia> c = 1 + BigFloat(-pi, precision=32)*im
1.0 - 3.14159265346825122833251953125im

julia> precision(imag(d))
32

julia> precision(imag(c))
32
```

### PR DESCRIPTION
**Pull request by AshlinHarris**
Fixes #52100

Currently, [`-(::BigFloat)`](https://github.com/JuliaLang/julia/blob/187e8c2222878c68b2afc9295ab8dc61773bd7f2/base/mpfr.jl#L689C1-L693C4) returns a value with the default `BigFloat` precision. With this commit, `-(x::BigFloat)` is altered to return a value with the same precision as the input.

### CODE REVIEW

**Comment by AshlinHarris** [comment]
Example behavior in master branch:
```
julia> x = BigFloat(pi, precision=32);

julia> (x,-x)
(3.1415926535, -3.14159265346825122833251953125)
```

Example behavior in this commit:

```
julia> x = BigFloat(pi, precision=32);

julia> (x,-x)
(3.14159265346825122833251953125, -3.14159265346825122833251953125)
```

**Comment by AshlinHarris** [comment]
Now I think the issue is related to `precision`, not to `show`:
```
julia> a = BigFloat(pi, precision=32)
3.1415926535

julia> b = BigFloat(-pi, precision=32)
-3.1415926535

julia> c = -a
-3.14159265346825122833251953125

julia> d = -b
3.14159265346825122833251953125

julia> precision.([a,b,c,d])
4-element Vector{Int64}:
  32
  32
 256
 256

```

**Comment by AshlinHarris** [comment]
`-(x::BigFloat)` now uses `precision(x)` for the return value, as opposed to `precision(BigFloat)`. This seems to solve the issue, but I'm not sure if this could be a breaking change, or if other functions might need to be updated for consistency.

File: base/mpfr.jl [hunk 1]
```diff
@@ -687,9 +687,11 @@ for (fJ, fC, fI) in ((:+, :add, 0), (:*, :mul, 1))
 end
 
 function -(x::BigFloat)
-    z = BigFloat()
-    ccall((:mpfr_neg, libmpfr), Int32, (Ref{BigFloat}, Ref{BigFloat}, MPFRRoundingMode), z, x, ROUNDING_MODE[])
-    return z
+    setprecision(precision(x)) do
+        z = BigFloat()
```
Comment (KlausC) [inline base/mpfr.jl:691]: Why not simply `z = BigFloat(;precision=precision(x))` ? That would avoid changing the global setting.

Comment (AshlinHarris) [reply base/mpfr.jl:691]: No reason. `z = BigFloat(;precision=precision(x))` is more readable, but I hadn't thought of it at the time. But I don't think `setprecision` will change anything outside the `do` block.

Comment (KlausC) [reply base/mpfr.jl:691]: It is not thread-safe, though. That means during task switches during the `do` block, the global setting of one of the tasks could be undefined.

**Comment by oscardssmith** [comment]
I agree that this makes sense, but don't think we should do it since it is unlike how all other bigfloat operations work.

**Comment by AshlinHarris** [comment]
As far as I can tell, `-(::BigFloat)` is the only operation on one input where the user might expect precision to be carried over. The closest thing to another example I could find is [`modf(::BigFloat)`](https://github.com/JuliaLang/julia/blob/39ccdb27990d0095d44715394bf2cb18f6b14dd3/base/mpfr.jl#L826C1-L831C4), where a user might expect the fractional part to have the same number of digits, but that would still have to be calculated.

Here is the behavior of `modf` in base Julia:
```julia
julia> x = BigFloat(pi, precision=32)
3.1415926535

julia> modf(x)
(0.14159265346825122833251953125, 3.0)
```
We could try the same fix as in `-(::BigFloat)`:

```julia
function modf(x::BigFloat)
    zint = BigFloat()
    zfloat = BigFloat(; precision=_precision(x)) # Function was changed here
    ccall((:mpfr_modf, libmpfr), Int32, (Ref{BigFloat}, Ref{BigFloat}, Ref{BigFloat}, MPFRRoundingMode), zint, zfloat, x, ROUNDING_MODE[])
    return (zfloat, zint)
end
```

But using input precision doesn't resolve the issue:

```julia
julia> x = BigFloat(pi, precision=32)
3.1415926535

julia> modf(x)
(0.14159265347, 3.0)

```

For all the other functions, output precision could possibly be based on the input, but each would have to be calculated, so it makes sense to fall back to default precision in each case. So it might make sense to change the behavior of `-(BigFloat)` only, since this is the only function where the user would expect the output precision to exactly match the input.

**Comment by JeffBezanson** [comment]
Triage thinks this is too breaking. We could potentially do something else with bigfloat precision like using a ScopedValue for it, though.

**Comment by vtjnash** [comment]
What makes this breaking? Isn't this just happen to be one of the only cases where preserving the input precision is certain not to lose any bits?

**Comment by nsajko** [comment]
> What makes this breaking?

I don't know about the triage decision, but the precision is part of the the `BigFloat` interface, a user might call `precision` on a `BigFloat` value and the inconsistency could cause a bug.

**Comment by mbauman** [comment]
`abs` would also be a similar candidate.  Is loosing precision more compelling?

```julia
julia> x = nextfloat(BigFloat(-1, precision=512))
-0.999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999925

julia> -x
1.0

julia> abs(x)
1.0
```

**Comment by LilithHafner** [comment]
Always keep precision > never keep precision > sometimes keep precision

Folks at triage did say "people don't want a better answer, they want the same answer" which is reasonable. I think this (and changing precision of all ops to preserve precision) is a minor change, but nanosoldier is not going to give great results for that kind of breakage.

Still, @nanosoldier `runtests()`

**Comment by simonbyrne** [comment]
`ldexp` and `frexp` and `significand` are good candidates too

**Review by KlausC — COMMENTED** [review]
use `precision(x)` instead of  `_precision(x)`

File: base/mpfr.jl [hunk 2]
```diff
@@ -694,7 +694,7 @@ for (fJ, fC, fI) in ((:+, :add, 0), (:*, :mul, 1))
 end
 
 function -(x::BigFloat)
-    z = BigFloat()
+    z = BigFloat(; precision=_precision(x))
```
Comment (mbauman) [inline base/mpfr.jl:697]:
```suggestion
    z = BigFloat(; precision=precision(x))
```

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/52288#issuecomment-1971894880) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/5cac38d_vs_c06662a/report.html) is available.

**Comment by LilithHafner** [comment]
Nanosoldier results are garbage because I triggered nanosoldier before @mbauman fixed the bug reported by @KlausC  in https://github.com/JuliaLang/julia/pull/52288#pullrequestreview-1910715955

@nanosoldier `runtests()`

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/52288#issuecomment-1974813465) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/e0365eb_vs_c06662a/report.html) is available.

**Comment by LilithHafner** [comment]
Nanosoldier is clean. As I said previously, that doesn't mean much in this case. But it is a thing.

**Comment by mbauman** [comment]
Might be worth asking triage for a second look?  :)

We have a bit more context now — this _would be_ a minor change and it's not necessarily the bug fix for the printing issue.  There are currently two functions that preserve precision as far as I've found: `nextfloat` and `prevfloat`.

I think that `significand`, `frexp`, and `ldexp` should also be added to the list.  The common thread there is that they are all pretty close the floating point representation itself.

Then the interesting thing with unary minus is that `abs(x)` is effectively `x < 0 ? -x : x`.

These should probably all happen within the same release — and would require NEWS.  But I think they're all well-motivated.

**Comment by LilithHafner** [comment]
I concur. I think that before we discuss it on triage we should get nanosoldier results for making all those changes (together in one PR), just so we know for sure we're not totally totally breaking.

The "bugfix" label is technicality accurate in but misleading +1 for removing it.

**Comment by oscardssmith** [comment]
I definitely agree on `significand`, `frexp`, and `ldexp`. `abs` and `-` are on the edge to me, but I definitely agree that the proposed behavior is more useful.

**Comment by mbauman** [comment]
Ah, `copysign` is another `abs`-like function that relies on unary minus:

```Julia
julia> copysign(prevfloat(BigFloat(1.0; precision=512)), 1)
0.999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999925

julia> copysign(prevfloat(BigFloat(1.0; precision=512)), -1)
-1.0
```

**Comment by KlausC** [comment]
Bump

**Pull request closed without merging**